### DISCUSSION TITLE
How to overlay a flat image onto a 3D mesh?

### DISCUSSION THREAD

**Discussion by PeterJBassett** [iTwin/itwinjs-core#4883] [Q&A]
Hello I am working on the FutureOn FieldTwin to iTwin connector.
In FutureOn FieldTwin there is the ability to overlay an image onto a 3D mesh, so that the image appears 'draped' over the 3D surface. For example see below:
<img width="596" alt="Screenshot 2023-01-06 at 16 30 18" src="https://user-images.githubusercontent.com/61423714/211055843-104dae59-6927-40e7-80db-3b98a614b8b1.png">

My question is - is it possible to do this in an iModel using the iTwin.js library?
If so, please could you provide some guidance on how to do this using iTwin.js?

**Comment by pmconne** [comment]
Hi Peter,

iTwin.js uses texture draping for various purposes - mostly related to terrain and map imagery.
Can you please describe the problem that you want to solve by employing this technique?

Also, can you supply some more details:
- Are you draping the image onto a specific element? A specific model? The entire iModel? A reality mesh?
- How large an area does the mesh cover (meters? kilometers? etc?)
- Is the draped image part of the iModel/element, or is it something that is expected to be applied (and possibly removed or changed) interactively at display time?
- I assume you considered using an ordinary textured material. In what ways does that fail to solve the problem in this case?

**Reply by PeterJBassett** [reply]
Hello Paul @pmconne 
Thank you for your reply. 
I am still not much of an expert with the iTwin.js library and am not completely familiar with all the concepts and terms.
As shown in the image above, I have a requirement in the FieldTwin to iModel connector to be able to pass a jpg image file from FieldTwin to the iModel eg. the blue square image shown, and drape that over a mesh of 3D points as shown.
To answer your specific questions:
* Yes this is just an element within an iModel, the iModel will contains a lot of other 3d elements.
* The mesh could be any size from meters to kilometers.
* The image is a static jpg file, it can only be updated by re-running the connector synchronisation.
* In the existing connector I have an example where a texture is tiled over a large 3d mesh. However, in this case I don't want to tile the texture over the mesh, I just want a single instance of the jpg image which has the mesh shape applied to it.

Do you know of any example code on the iTwin website that show how to do this type of 3d rendering with images and meshes, that I can look at to get a better understanding of the library?

Thanks for your help.

**Reply by pmconne** [reply]
Thanks @PeterJBassett. One other factor that will determine how much effort this would require: is the grey in your screenshot the natural color of the mesh? Is it necessary to use the mesh's natural color outside of the draped area, or would it suffice to have the material specify that pixels outside of the draped area should be a specific color (e.g., grey) that may not be the same as the mesh's natural color?

**Comment by PeterJBassett** [comment]
Hello @pmconne 
Thanks for your reply, the grey mesh shown is just there to show the underlying 3d mesh that the image should be 'draped' over.
This exists as a separate element in the iModel.

What I want to be able to do is apply the 3d mesh shape to the image itself. So it would only be the mesh points that are within the image x,y boundary that are used to make the flat 2d image into a 3d shape.

**Reply by PeterJBassett** [reply]
So just showing the image only without showing the grey mesh.
It would look like this, with the image taking the shape of the 3d mesh.

<img width="405" alt="Screenshot 2023-01-19 at 11 36 10" src="https://user-images.githubusercontent.com/61423714/213432695-9867c88a-c042-49bb-b6e0-07e481442d59.png">

**Reply by pmconne** [reply]
Thanks for elaborating. So you've got two elements, one of which is a rectangular image to be draped onto the shape of the other, and you may want to hide the 3d element.
Can I ask what Futureon users use this feature for?

**Comment by PeterJBassett** [comment]
Hi @pmconne It is just the image that I need to know how to process. I just want to know how to drape it onto a 3d mesh.
The grey 3d surface is already handled ok by the connector.

I am hoping that this can be done with the existing itwinjs libraries, I just need some guidance on which itwin modules to use, and any pointers to examples on the website showing how to use the libraries.

I am not an oil and gas systems expert, but my understanding is that the sea floor bathymetry is modelled as a 3d mesh, and the users want to overlay an image onto the sea-floor, this might be a map of the best drilling locations for example.

**Reply by pmconne** [reply]
Right. We [already have support](https://www.itwinjs.org/reference/core-common/maplayers/modelmaplayersettings/) for draping the contents of any number of models onto the 3d global terrain. If the seafloor geometry was being obtained from a terrain server, your problem would be solved.
We do not currently support draping onto elements like your seafloor mesh.

**Reply by PeterJBassett** [reply]
Hello @pmconne Ok, thanks for your reply, in that case, I will look at some simplified solution for the user.
Thanks anyway for all your help.
